flume-ng-sinks/flume-ng-elasticsearch-sink/src/main/java/org/apache/flume/sink/elasticsearch/client/ElasticSearchRestClient.java

```diff
@@ -105,7 +105,7 @@ public void addEvent(Event event, IndexNameBuilder indexNameBuilder, String inde
     parameters.put(INDEX_OPERATION_NAME, indexParameters);
 
     Gson gson = new Gson();
-    synchronized (bulkBuilder) {
```

Comment:
Note to reviewers: SpotBugs was reporting `ML_SYNC_ON_FIELD_TO_GUARD_CHANGING_THAT_FIELD`.

Reply:
Looking at this PR, all changes seem awesome to me, but this one (and the one after - same `synchronized (this)`) seems to alter semantics. What do you think about going one step further and making `bulkBuilder` a `StringBuffer` instead of a `StringBuilder` so that the thread-safety of `StringBuffer` is leveraged?

Reply:
The way I understand it, switching from `synchronized()` to a thread-safe `StringBuffer` would create a bug.

There are 1+ "producer" threads that perform this:

```
synchronized (lock) {
  bulkBuilder.append(gson.toJson(parameters));
  bulkBuilder.append("\n");
  bulkBuilder.append(content.toBytesArray().toUtf8());
  bulkBuilder.append("\n");
}
```

And there is one "consumer" thread doing this:

```
synchronized (lock) {
  entity = bulkBuilder.toString();
  bulkBuilder = new StringBuilder();
}
```

I thought the reason we use  `synchronized()` here is to prevent the "consumer" thread from capturing an incomplete sequence of the four calls to `.append()`.  Either we get all four `.append()` calls or we get none of them.  Otherwise, the bulk insert request would be invalid.

In addition, you get the additional performance hit of synchronizing on each call to `.append()` -- I have not profiled, so I have no idea if this makes a difference in practice.

If `synchronized (this)` is an issue, maybe can switch to a dedicated lock object?

Reply:
If the `append`s happen in one call, there should be no race condition. Similarly, a `.setLength(0);` to empty the buffer should also allow for no race condition. Let me know if I've missed something. A second best option would be a dedicated lock object, IMHO.